File: cpp/547_省份数量_medium.cpp
class Solution {
public:
    int findCircleNum(vector<vector<int>>& isConnected) {
        int N = isConnected.size();
        vector<int> color(N, 0);
        int num = 0;
        for(int i=0; i<N; i++){
            if(color[i] == 0){
                num++;
                circleDFS(isConnected, color, i, -1);
            }
        }
        return num;
    }
    void circleDFS(vector<vector<int>> &isConnected, vector<int> &color, int v, int p){
        color[v] = 1;
        for(int i=0; i<isConnected.size(); i++){
            if(i==v || i==p)
                continue;
            if(isConnected[v][i] == 1 && color[i] == 0)
                circleDFS(isConnected, color, i, v);
        }
        color[v] = 2;
    }
};

File: python/547_省份数量_medium.py
class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        N = len(isConnected)
        color = [0 for i in range(N)]
        def circleDFS(color, v, p):
            color[v] = 1
            for i in range(N):
                if i==v or i==p:
                    continue
                if isConnected[v][i] == 1 and color[i] == 0:
                    circleDFS(color, i, v)
            color[v] = 2

        num = 0
        for i in range(N):
            if color[i] == 0:
                num += 1
                circleDFS(color, i, -1)
        return num
    